Release 2.4

jfy133 opened the pull request:
See changelog: https://github.com/nf-core/mag/blob/dev/CHANGELOG.md#240---2023-09-07

Note large number of files is due to template updates!

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/mag/tree/master/.github/CONTRIBUTING.md)- [ ] If necessary, also make a PR on the nf-core/mag _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

d4straub commented on docs/output.md:
```diff
@@ -377,6 +425,22 @@ By default, only the raw bins (and unbinned contigs) from the actual binning met
 
 ⚠️ Due to ability to perform downstream QC of both raw and refined bins in parallel (via `--postbinning_input)`, bin names in DAS Tools's `*_allBins.eval` file will include `Refined`. However for this particular file, they _actually_ refer to the 'raw' input bins. The pipeline renames the input files prior to running DASTool to ensure they can be disambiguated from the original bin files in the d ...
 
+### Tiara
```
thats not rendered properly, see above https://nf-co.re/mag/dev/docs/output/#bin-sequencing-depth

d4straub commented on docs/usage.md:
````diff
@@ -85,7 +137,30 @@ work                # Directory containing the nextflow working files
+```yaml
+input: './samplesheet.csv'
+outdir: './results/'
+input: 'data'
````
twice input? that might break...
```suggestion
```

d4straub commented on docs/usage.md:
```diff
@@ -112,6 +191,8 @@ To allow also reproducible bin QC with BUSCO, run BUSCO providing already downlo
 
 For the taxonomic bin classification with [CAT](https://github.com/dutilh/CAT), when running the pipeline with `--cat_db_generate` the parameter `--save_cat_db` can be used to also save the generated database to allow reproducibility in future runs. Note that when specifying a pre-built database with `--cat_db`, currently the database can not be saved.
 
+When it comes to visualizing taxonomic data using [Krona](https://github.com/marbl/Krona), you have the option to provide a taxonomy file, such as `taxonomy.tab`, using the `--krona_db` parameter. If you don't supply a taxonomy file, Krona is designed to automatically download the required taxonomy data for visualization. If you choose to provide a pre-existing taxonomy file using the `--krona_db ...
```
The last 2 sentences seem repetitive of the first two? Maybe just me

jfy133 replied:
Agreed it is

d4straub commented on modules/local/centrifuge.nf:
```diff
@@ -13,18 +13,20 @@ process CENTRIFUGE {
     tuple val("centrifuge"), val(meta), path("results.krona"), emit: results_for_krona
     path "report.txt"                                        , emit: report
-    path "kreport.txt"                                       , emit: kreport
+    tuple val(meta), path("*kreport.txt")                     , emit: kreport
```
Seems like indentation isnt following the 'Harshil Alignment Format'

d4straub commented on modules/nf-core/genomad/download/genomad-download.diff:
```diff
@@ -0,0 +1,14 @@
+     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+         'https://depot.galaxyproject.org/singularity/genomad:1.5.2--pyhdfd78af_0':
+-        'biocontainers/genomad:1.5.2--pyhdfd78af_0' }"
++        'quay.io/biocontainers/genomad:1.5.2--pyhdfd78af_0' }"
```
So that isnt updated yet in the pipeline? I thought this is supposed to be not handled by `singularity.registry`  ?

jfy133 replied:
@CarsonJM @PhilPalmer can uo comment on this? Any reason why the patch reverting?

CarsonJM replied:
For some reason when I was adding this module, the tests were not passing until I made the patch. Here was the comment you left prior to merging (https://github.com/nf-core/mag/pull/364#discussion_r1255350233)

jfy133 replied:
Aha! Well future James forgot pastbJames. Thanks for the reminder!

jfy133 replied:
Aha! Well future James forgot past James. Thanks for the reminder!

CarsonJM replied:
@jfy133 Is there anything you would like me to do to help with this PR? I am happy to review it and help make changes if needed!

jfy133 replied:
Sorry for the delay, short break form parentla leave for a conference... so no difference in time availablility :sweat_smile: 

I think we are good for now @CarsonJM thank you! the changes in the comments are not too complicated.
 It will be monitoring issues/slack reports for bugs when the release is out (and then rapid subsequent releases after) that will be where help would benefit most <3

CarsonJM replied:
I'm happy to help where I can! Sorry I have been less responsive lately, I have my parental leave coming up soon so I have been frantically trying to finish up all my wet-lab work before then! 😅

d4straub commented on modules/nf-core/metaeuk/easypredict/main.nf:
```diff
@@ -0,0 +1,62 @@
+    conda "bioconda::metaeuk=6.a5d39d9"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/metaeuk:6.a5d39d9--pl5321hf1761c0_2':
+        'biocontainers/metaeuk:6.a5d39d9--pl5321hf1761c0_2' }"
```
Hm but e.g. here `quay.io` isnt added, will that mixture work fine? Maybe I am missing something.

jfy133 replied:
I guess I need to do a pass through all modules, they've not been updated with these changes yet (as you commented on slack)

jfy133 replied:
Done!

d4straub commented on subworkflows/local/depths.nf:
```diff
@@ -0,0 +1,65 @@
+params.mag_depths_options                           = [:]
+params.mag_depths_plot_options                      = [:]
+params.mag_depths_summary_options                   = [:]
```
what are those for, seem not be used?

jfy133 replied:
I will have to check, that was carried over from the original code I imagine.

jfy133 replied:
Don't find any reference to them anywhere else :+1:

ramprasadn commented on conf/modules.config:
```diff
@@ -197,6 +232,7 @@ process {
     }
 
     withName: KRAKEN2 {
+        ext.prefix = { ${meta.id} }
```
This is same as the prefix assigned by the module, right? Did you mean to assign a different prefix?

jfy133 replied:
I think it was a copy paste error :grimacing:

ramprasadn commented on conf/modules.config:
```diff
@@ -189,6 +223,7 @@ process {
     }
 
     withName: CENTRIFUGE {
+        ext.prefix = { ${meta.id} }
```
This is same as the prefix assigned by the module, right? Did you mean to assign a different prefix?

jfy133 replied:
See above!

d4straub commented on workflows/mag.nf:
```diff
@@ -320,48 +414,57 @@ workflow MAG {
-        ch_long_reads = PORECHOP.out.reads
-        ch_versions = ch_versions.mix(PORECHOP.out.versions.first())
-    }
+                        .map {
```
indentation seems off

jfy133 replied:
Fixed

ramprasadn commented on conf/modules.config:
```diff
@@ -252,29 +288,42 @@ process {
+                mode: params.publish_dir_mode,
+                pattern: "*.{bam,bai}",
+                enabled: params.save_assembly_mapped_reads
+            ],
```
Didn't know we could pass two or more rules to publishDir like this. Pretty cool!

jfy133 replied:
Nifty isn't it!

d4straub reviewed (commented):
Huge PR, I am sure I missed points.
I am mostly concerned about the container registry, I made a similar mix in ampliseq release 2.6.0 and had to follow up with 2.6.1 real quick because it didnt work on AWS (local/cluster passed if I remember correctly).

ramprasadn commented on subworkflows/local/binning.nf:
```diff
@@ -36,17 +25,16 @@ workflow BINNING {
     ch_summarizedepth_input = assemblies
                                 .map { meta, assembly, bams, bais ->
-                                        def meta_new = meta.clone()
+                                        def meta_keys = meta.keySet()
+                                        def meta_new = meta + meta.subMap(meta_keys)
```
Wouldn't `meta_new` be the same as `meta`? 🤔

jfy133 replied:
Ehhh, yes :sweat_smile:

ramprasadn commented on subworkflows/local/binning.nf:
```diff
@@ -136,63 +120,20 @@ workflow BINNING {
     GUNZIP_BINS ( ch_final_bins_for_gunzip )
     ch_binning_results_gunzipped = GUNZIP_BINS.out.gunzip
-    ch_versions = ch_versions.mix(GUNZIP_BINS.out.versions.first())
+        .groupTuple(by: 0)
```
Do you think using `size` or `groupKey` will make this grouping operation more efficient?

jfy133 replied:
Possibly, but it's not trivial as the number of bins files vary per sample. I could upstream calculate the number of bins, insert that into a meta field for the sample, and then insert that into `size`, but this would require experimentation. I've made an issue (tagged you) to investigate for a future release, as I don't think that's a blocker at the moment.

ramprasadn commented on subworkflows/local/binning.nf:
```diff
@@ -136,63 +120,20 @@ workflow BINNING {
-    // create file containing group information for all samples
-    ch_sample_groups = reads
-        .collectFile(name:'sample_groups.tsv'){ meta, reads -> meta.id + '\t' + meta.group + '\n' }
+        .groupTuple(by: 0)
```
Do you think using `size` or `groupKey`will make this grouping operation more efficient?

jfy133 replied:
see https://github.com/nf-core/mag/pull/499#discussion_r1326993376

ramprasadn commented on subworkflows/local/depths.nf:
```diff
@@ -0,0 +1,65 @@
+            [ meta_new, bins, contig_depths_file ]
+        }
+        .transpose()
+        .groupTuple(by: [0,2])
```
Do you think using `size` or `groupKey` will make grouping operation more efficient here?

jfy133 replied:
see https://github.com/nf-core/mag/pull/499#discussion_r1326993376

ramprasadn commented on workflows/mag.nf:
```diff
@@ -96,10 +94,13 @@ include { BINNING_PREPARATION } from '../subworkflows/local/binning_preparation'
 include { BINNING             } from '../subworkflows/local/binning'
 include { BINNING_REFINEMENT  } from '../subworkflows/local/binning_refinement'
 include { BUSCO_QC            } from '../subworkflows/local/busco_qc'
+include { VIRUS_IDENTIFICATION} from '../subworkflows/local/virus_identification'
 include { CHECKM_QC           } from '../subworkflows/local/checkm_qc'
 include { GUNC_QC             } from '../subworkflows/local/gunc_qc'
 include { GTDBTK              } from '../subworkflows/local/gtdbtk'
 include { ANCIENT_DNA_ASSEMBLY_VALIDATION } from '../subworkflows/local/ancient_dna'
+include { DOMAIN_CLASSIFICATION } from '../subworkflows/local/domain_classification'
+include { DEPTHS              } from '../subworkflows/local/depths'
```
Curly brackets are not aligned here (and I am only pointing it out since they are in alignment everywhere else 😆)

jfy133 replied:
Done!

jfy133 reviewed (commented):
Responses from @ramprasadn 's rfist set of comments

jfy133 commented on docs/images/mag_workflow.svg:
I'm not so sure about that - I find it's a good reference to know which version has what functionality - and also nice for making 'progression' animations ;) - even if not regular!

It's a very easy fix to update it in each one (Can even be done programically to be honest). I can take responsibility for it ;)

jfy133 commented on README.md:
```diff
@@ -36,46 +33,41 @@ The pipeline then:
 - performs metagenome binning using [MetaBAT2](https://bitbucket.org/berkeleylab/metabat/src/master/), [MaxBin2](https://sourceforge.net/projects/maxbin2/), and/or with [CONCOCT](https://github.com/BinPro/CONCOCT), and checks the quality of the genome bins using [Busco](https://busco.ezlab.org/), or [CheckM](https://ecogenomics.github.io/CheckM/), and optionally [GUNC](https://grp-bork.embl-commu ...
 - optionally refines bins with [DAS Tool](https://github.com/cmks/DAS_Tool)
-- assigns taxonomy to bins using [GTDB-Tk](https://github.com/Ecogenomics/GTDBTk) and/or [CAT](https://github.com/dutilh/CAT)
+- assigns taxonomy to bins using [GTDB-Tk](https://github.com/Ecogenomics/GTDBTk) and/or [CAT](https://github.com/dutilh/CAT) and optionally identifies viruses in assemblies using [geNomad](https://github.com/apcamargo/genomad)
```
My understadning this list was originally designed for just default run, but it's gone much larger. I've added a few more things too

jfy133 commented:
@ramprasadn I believe I've addressed all of your comments now.

I have one more comment from Daniel about updating the URLs in the modules, but hopefully you are happy with everything else!

d4straub commented on conf/modules.config:
```diff
@@ -409,7 +456,7 @@ process {
     }
 
-    withName: GTDBTK_CLASSIFY {
+    withName: GTDBTK_CLASSIFYWF {
```
The name addition `WF` seems here out of place. Is that correct?

jfy133 replied:
That's called what it is in the module I think?

d4straub replied:
Indeed, weird naming for a module.

d4straub reviewed (approved):
LGTM.

However, the template update came at a bad time...

jfy133 commented:
I am going for a release (merging despite the linting error), and doing an immediate followup release PR with the template
